mods/timer-resolution-control.wh.cpp

```diff
@@ -215,25 +221,84 @@ BOOL Wh_ModInit(void)
+
     LoadSettings();
 
     Wh_SetFunctionHook((void*)pNtSetTimerResolution, (void*)NtSetTimerResolutionHook, (void**)&pOriginalNtSetTimerResolution);
 
     return TRUE;
 }
 
-void Wh_ModSettingsChanged(void)
+void EnforceLimits() 
+{
+    ULONG CurrentResolution;
+    NTSTATUS status = pOriginalNtSetTimerResolution(0, FALSE, &CurrentResolution);
+    if (status == STATUS_TIMER_RESOLUTION_NOT_SET) {
+        Wh_Log(L"NtSetTimerResolution has not been called by the program");
+        status = STATUS_SUCCESS;
+        CurrentResolution = g_lastDesiredResolution;    //Take the value obtained from NtQueryTimerResolution, which does not fail with STATUS_TIMER_RESOLUTION_NOT_SET. Even if the program has not set the resolution, I have observed that the system itself may have set the resolution on program start to 1ms for some reason.
+    }
+    if (NT_SUCCESS(status)) {
+        Wh_Log(L"NtGetTimerResolution: current=%f", (double)CurrentResolution / 10000.0);
+
+        ULONG limitResolution = g_limitResolution;
+        ULONG lastDesiredResolution = g_lastDesiredResolution;
+        if (CurrentResolution < limitResolution
+            || CurrentResolution < lastDesiredResolution) {
+
+            Wh_Log(L"> CurrentResolution: %f milliseconds", (double)CurrentResolution / 10000.0);
+
+            if (lastDesiredResolution < limitResolution) {
+                Wh_Log(L"* Overriding resolution: %f milliseconds", (double)limitResolution / 10000.0);
+                pOriginalNtSetTimerResolution(limitResolution, TRUE, &CurrentResolution);
+            }
+            else {
+                Wh_Log(L"* Restoring resolution: %f milliseconds", (double)lastDesiredResolution / 10000.0);
+                pOriginalNtSetTimerResolution(lastDesiredResolution, TRUE, &CurrentResolution);
+            }
+        }
+    }
+    else {
+        Wh_Log(L"NtGetTimerResolution failed with status: 0x%X", status);
+    }
+}
+
+void Wh_ModAfterInit(void) 
+{  
+    //Force mod timer resolution settings if the resolution was already changed by the program. NB! In order to avoid any race conditions, do this only after the hook is activated. Else there might be a situation that the mod overrides the current resolution while hook is not yet set, and then the program changes it again before the hook is finally set.
+
+    EnforceLimits();
+}
+
+void Wh_ModSettingsChanged(void) 
 {
     Wh_Log(L"SettingsChanged");
 
     LoadSettings();
+
+    EnforceLimits();
 }
+
+void Wh_ModUninit() 
+{
+    Wh_Log(L"Uniniting...");
+
+    //Lift all limits and restore original resolution set by the program
+    ULONG lastDesiredResolution = g_lastDesiredResolution;
+    Wh_Log(L"Restoring desired resolution: %f milliseconds", (double)lastDesiredResolution / 10000.0);
+    ULONG CurrentResolution;
+    pOriginalNtSetTimerResolution(g_lastDesiredResolution, TRUE, &CurrentResolution);
```

Comment:
Using `pOriginalNtSetTimerResolution` here is unsupported since the hooks are already removed. It's a tricky one, it coincidentally works when the mod is disabled, but it will likely crash when you exit Windhawk. Just use `NtSetTimerResolution` instead.